sdks/python/apache_beam/dataframe/transforms_test.py

```diff
@@ -348,6 +350,44 @@ def test_rename(self):
               }, errors='raise'))
 
 
+class FusionTest(unittest.TestCase):
+  @staticmethod
+  def fused_stages(p):
+    return p.result.metrics().query(
+        metrics.MetricsFilter().with_name(
+            fn_runner.FnApiRunner.NUM_FUSED_STAGES_COUNTER)
+    )['counters'][0].result
```

Comment:
It feels a little odd to verify this through a metric in `FnApiRunner`. Could we instrument DataFrameTransform expansion isntead?

Reply:
I'm open to ideas here. At the end of the day, I'm wanting to assert that things get sufficient fused (possibly catching issues with side inputs as well as shuffles).

Reply:
The other thought I had was just traversing the pipeline and counting CoGBKs, but you're right that wouldn't help if side inputs mess up fusion.

Could we pull out  the fusion logic?